File: app/src/main/java/ml/sky233/suiteki/service/ble/FirmwareInstaller.java
package ml.sky233.suiteki.service.ble;

import android.annotation.SuppressLint;
import android.bluetooth.BluetoothGatt;
import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.util.Log;

import com.clj.fastble.BleManager;
import com.clj.fastble.callback.BleWriteCallback;
import com.clj.fastble.data.BleDevice;
import com.clj.fastble.exception.BleException;

import java.util.ArrayList;


import ml.sky233.suiteki.MainApplication;
import ml.sky233.suiteki.callback.ProgressCallback;
import ml.sky233.suiteki.util.BleLogTools;
import ml.sky233.suiteki.util.BytesUtils;
import ml.sky233.suiteki.util.MsgUtils;

@SuppressLint("MissingPermission")
public class FirmwareInstaller {

    public static byte[] d0 = {-48}, d1 = {-47}, d2 = {}, d3 = {-45, 1}, d5 = {-43}, d6 = {-42};
    public ArrayList<byte[]> doPerformBytes = new ArrayList<>(), finalPerformBytes = new ArrayList<>();
    int doPerform = 0;
    int finalPerform = 0;
    int dataIndex = 0;
    Context context;
    public BluetoothGatt mGatt;
    public byte[] firmware_bytes;
    public ArrayList<byte[]> spilt_byte;
    BleDevice bleDevice;
    Handler handler;
    ProgressCallback callback;

    public FirmwareInstaller(Context context, BluetoothGatt gatt, byte[] bytes, BleDevice bleDevice, Handler handler) {
        this.mGatt = gatt;
        this.firmware_bytes = bytes;
        this.bleDevice = bleDevice;
        this.context = context;
        this.handler = handler;
        BleService.setStatus(HuamiService.STATUS_BLE_INSTALLING);
        spilt_byte = BytesUtils.spiltBytes(bytes);
        d2 = BytesUtils.getD2(bytes, HuamiService.TYPE_D2_WATCHFACE);
        doPerformBytes.add(d0);
        doPerformBytes.add(d1);
        doPerformBytes.add(d2);
        doPerformBytes.add(d3);
        finalPerformBytes.add(d5);
        finalPerformBytes.add(d6);
        doPerform();
    }

    public FirmwareInstaller(Context context, BluetoothGatt gatt, byte[] bytes, BleDevice bleDevice, Handler handler, ProgressCallback callback) {
        this.mGatt = gatt;
        this.firmware_bytes = bytes;
        this.bleDevice = bleDevice;
        this.context = context;
        this.handler = handler;
        this.callback = callback;
        BleService.setStatus(HuamiService.STATUS_BLE_INSTALLING);
        spilt_byte = BytesUtils.spiltBytes(bytes);
        d2 = BytesUtils.getD2(bytes, HuamiService.TYPE_D2_WATCHFACE);
        doPerformBytes.add(d0);
        doPerformBytes.add(d1);
        doPerformBytes.add(d2);
        doPerformBytes.add(d3);
        finalPerformBytes.add(d5);
        finalPerformBytes.add(d6);
        doPerform();
    }

    public void setStatus(int status) {
        handler.sendMessage(MsgUtils.build(status, 1));
    }

    public void writeFirmware(byte[] bytes) {
        BleManager.getInstance().write(bleDevice,
                HuamiService.UUID_SERVICE_FIRMWARE.toString(),
                HuamiService.UUID_CHARACTERISTIC_FIRMWARE_WRITE.toString(),
                bytes, new BleWriteCallback() {
                    @Override
                    public void onWriteSuccess(int i, int i1, byte[] bytes) {
                        BleLogTools.onWriteSuccess(bytes);
                    }

                    @Override
                    public void onWriteFailure(BleException e) {
                        BleLogTools.onWriteFailure(bytes);
                    }
                });
        try {
            Thread.sleep(500);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    public void writeNotify(byte[] bytes) {
        BleManager.getInstance().write(bleDevice,
                HuamiService.UUID_SERVICE_FIRMWARE.toString(),
                HuamiService.UUID_CHARACTERISTIC_FIRMWARE_NOTIFY.toString(),
                bytes, new BleWriteCallback() {
                    @Override
                    public void onWriteSuccess(int i, int i1, byte[] bytes) {
                        BleLogTools.onWriteSuccess(bytes);
                    }

                    @Override
                    public void onWriteFailure(BleException e) {
                        BleLogTools.onWriteFailure(bytes);
                    }
                });
    }

    public void doPerform() {
        if (doPerform >= doPerformBytes.size()) return;
        writeNotify(doPerformBytes.get(doPerform));
        doPerform++;
    }

    public void install() {
        dataIndex = 0;
        writeBytes();
    }

    public void writeBytes() {
        new Thread(() -> {
            int remaining = Math.min(spilt_byte.size() - dataIndex, 34);
            for (int i = 0; i < remaining; i++) {
                if (callback != null)
                    callback.onProgressChange((int) (((double) dataIndex / (double) spilt_byte.size()) * (double) 100));
                else
                    context.sendBroadcast(new Intent()
                            .setAction(BleActions.ACTION_BLE_FIRMWARE_INSTALLING)
                            .putExtra("progress", (int) (((double) dataIndex / (double) spilt_byte.size()) * (double) 100)));
                byte[] data = spilt_byte.get(dataIndex);
                dataIndex++;
                writeFirmware(data);
            }
        }).start();
    }

    public void onCharacteristicChange(Intent intent) {
        Log.d(MainApplication.TAG, intent.getAction());
        if (intent.getAction().equals(BleActions.ACTION_BLE_FIRMWARE_NOTIFY)) {
            byte[] value = intent.getByteArrayExtra("value");
            switch (BytesUtils.bytesToHexStr(value)) {
                case "10D001050020"://D0
                case "10D10100"://D1
                case "10D2010000000000000000"://D2
                    doPerform();
                    break;
                case "10D203"://D2
                    MainApplication.mService.reAuth(() -> {
                        doPerform = 0;
                        doPerform();
                    });
                    return;
                case "10D347":
                    //空间不足
                    setStatus(HuamiService.STATUS_BLE_NORMAL);
                    context.sendBroadcast(new Intent()
                            .setAction(BleActions.ACTION_BLE_FIRMWARE_INSTALL_FAILURE));
                    return;
                case "10D301"://D3
                    install();
                    return;
                case "10D501"://D5
                    finalPerform();
                    return;
                case "10D601"://D6
                    //安装完毕
                    setStatus(HuamiService.STATUS_BLE_NORMAL);
                    context.sendBroadcast(new Intent()
                            .setAction(BleActions.ACTION_BLE_FIRMWARE_INSTALLED));
                    return;
                default:
                    if (value[0] == 16 && value[1] == -44) {//10D4
                        if (dataIndex >= spilt_byte.size())
                            finalPerform();
                        else
                            writeBytes();
                        break;
                    }
                    if (value[0] == 16 && value[1] == -46 && value[2] == 1) {
                        doPerform();
                        break;//10D201
                    }
                    //安装失败
                    return;
            }
        }
    }

    public void finalPerform() {
        if (finalPerform >= finalPerformBytes.size()) return;
        writeNotify(finalPerformBytes.get(finalPerform));
        finalPerform++;
    }

}

File: app/src/main/java/ml/sky233/suiteki/service/ble/FirmwareInstaller.kt
package ml.sky233.suiteki.service.ble

import android.annotation.SuppressLint
import android.bluetooth.BluetoothGatt
import android.content.Context
import android.content.Intent
import android.os.Handler
import com.clj.fastble.BleManager
import com.clj.fastble.callback.BleWriteCallback
import com.clj.fastble.data.BleDevice
import com.clj.fastble.exception.BleException
import ml.sky233.suiteki.MainApplication
import ml.sky233.suiteki.MainApplication.Companion.context
import ml.sky233.suiteki.callback.ProgressCallback
import ml.sky233.suiteki.callback.ReAuthCallback
import ml.sky233.suiteki.util.BytesUtils
import ml.sky233.suiteki.util.MsgUtils

@SuppressLint("MissingPermission")
class FirmwareInstaller : ReAuthCallback {
    private var doPerformBytes = ArrayList<ByteArray>()
    private var finalPerformBytes = ArrayList<ByteArray>()
    private var doPerform = 0
    private var finalPerform = 0
    private var dataIndex = 0
    private var dataIndex_for_progress = 0 //该变量的定义只是为了进度条而存在
    private var mGatt: BluetoothGatt
    private var firmware_bytes: ByteArray
    private var spilt_byte: ArrayList<ByteArray>
    private var bleDevice: BleDevice
    private var handler: Handler
    var callback: ProgressCallback? = null

    constructor(
        gatt: BluetoothGatt,
        bytes: ByteArray,
        bleDevice: BleDevice,
        handler: Handler,
    ) {
        mGatt = gatt
        firmware_bytes = bytes
        this.bleDevice = bleDevice
        this.handler = handler
        BleService.Companion.setStatus(HuamiService.STATUS_BLE_INSTALLING)
        spilt_byte = BytesUtils.spiltBytes(bytes)
        d2 = BytesUtils.getD2(bytes, HuamiService.TYPE_D2_WATCHFACE)
        doPerformBytes.add(d0)
        doPerformBytes.add(d1)
        doPerformBytes.add(d2)
        doPerformBytes.add(d3)
        finalPerformBytes.add(d5)
        finalPerformBytes.add(d6)
        doPerform()
    }

    constructor(
        gatt: BluetoothGatt,
        bytes: ByteArray,
        bleDevice: BleDevice,
        handler: Handler,
        callback: ProgressCallback,
    ) {
        mGatt = gatt
        firmware_bytes = bytes
        this.bleDevice = bleDevice
        this.handler = handler
        this.callback = callback
        BleService.Companion.setStatus(HuamiService.STATUS_BLE_INSTALLING)
        spilt_byte = BytesUtils.spiltBytes(bytes)
        d2 = BytesUtils.getD2(bytes, HuamiService.TYPE_D2_WATCHFACE)
        doPerformBytes.add(d0)
        doPerformBytes.add(d1)
        doPerformBytes.add(d2)
        doPerformBytes.add(d3)
        finalPerformBytes.add(d5)
        finalPerformBytes.add(d6)
        doPerform()
    }

    fun setStatus(status: Int) {
        handler.sendMessage(MsgUtils.build(status, 1))
    }

    fun writeFirmware(bytes: ByteArray?) {
        BleManager.getInstance().write(bleDevice,
            HuamiService.UUID_SERVICE_FIRMWARE.toString(),
            HuamiService.UUID_CHARACTERISTIC_FIRMWARE_WRITE.toString(),
            bytes,
            true,
            true,
            0,
            object : BleWriteCallback() {
                override fun onWriteSuccess(i: Int, i1: Int, bytes: ByteArray) {
                    updateProgress()
                }

                override fun onWriteFailure(e: BleException) {
                }
            })
        try {
            Thread.sleep(200)
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
    }

    fun writeNotify(bytes: ByteArray?) {
        BleManager.getInstance().write(bleDevice,
            HuamiService.UUID_SERVICE_FIRMWARE.toString(),
            HuamiService.UUID_CHARACTERISTIC_FIRMWARE_NOTIFY.toString(),
            bytes,
            object : BleWriteCallback() {
                override fun onWriteSuccess(i: Int, i1: Int, bytes: ByteArray) {
                }

                override fun onWriteFailure(e: BleException) {
                }
            })
    }

    fun doPerform() {
        if (doPerform >= doPerformBytes.size) return
        writeNotify(doPerformBytes[doPerform])
        doPerform++
    }

    fun install() {
        dataIndex = 0
        writeBytes()
    }

    fun writeBytes() {
        val remaining = Math.min(spilt_byte.size - dataIndex, 34)
        var bytes = ByteArray(0)
        for (i in 0 until remaining) {
            bytes = BytesUtils.addTwoArray(bytes, spilt_byte[dataIndex])
            dataIndex++
        }
        writeFirmware(bytes)
    }

    private fun updateProgress() {
        if (callback != null) callback!!.onProgressChange((dataIndex_for_progress.toDouble() / spilt_byte.size.toDouble() * 100.0).toInt()) else context.sendBroadcast(
            Intent().setAction(BleActions.ACTION_BLE_FIRMWARE_INSTALLING).putExtra(
                "progress",
                (dataIndex_for_progress.toDouble() / spilt_byte.size.toDouble() * 100.0).toInt()
            )
        )
        dataIndex_for_progress++
    }

    fun onCharacteristicChange(intent: Intent) {

        if (intent.getAction() == BleActions.ACTION_BLE_FIRMWARE_NOTIFY) {
            val value: ByteArray = intent.getByteArrayExtra("value")!!
            when (BytesUtils.bytesToHexStr(value)) {
                "10D001050020", "10D00105002001", "10D10100", "10D2010000000000000000" -> doPerform()
                "10D203" -> {
                    MainApplication.mService?.reAuth(this)
                    return
                }

                "10D347" -> {
                    //空间不足
                    setStatus(HuamiService.STATUS_BLE_NORMAL)
                    context.sendBroadcast(
                        Intent().setAction(BleActions.ACTION_BLE_FIRMWARE_INSTALL_FAILURE)
                    )
                    return
                }

                "10D301" -> {
                    install()
                    return
                }

                "10D501" -> {
                    finalPerform()
                    return
                }

                "10D601" -> {
                    //安装完毕
                    setStatus(HuamiService.STATUS_BLE_NORMAL)
                    context.sendBroadcast(
                        Intent().setAction(BleActions.ACTION_BLE_FIRMWARE_INSTALLED)
                    )
                    return
                }

                else -> {
                    if (value[0].toInt() == 16 && value[1].toInt() == -44) { //10D4
                        if (dataIndex >= spilt_byte.size) finalPerform() else writeBytes()

                    }
                    if (value[0].toInt() == 16 && value[1].toInt() == -46 && value[2].toInt() == 1) {
                        doPerform()
                        //10D201
                    }
                    //安装失败
                    return
                }
            }
        }
    }

    fun finalPerform() {
        if (finalPerform >= finalPerformBytes.size) return
        writeNotify(finalPerformBytes[finalPerform])
        finalPerform++
    }

    companion object {
        var d0 = byteArrayOf(-48)
        var d1 = byteArrayOf(-47)
        var d2 = byteArrayOf()
        var d3 = byteArrayOf(-45, 1)
        var d5 = byteArrayOf(-43)
        var d6 = byteArrayOf(-42)
    }

    override fun reAuthSuccess() {
        doPerform = 0
        doPerform()
    }
}